### PR TITLE
Refactor the syntax file with many improvements

### PR DESCRIPTION
**Pull request by yuezk**
- Implements the JSX specification. https://github.com/facebook/jsx
- Syntax highlighting no longer dependents on pangloss/vim-javascript and yajs.vim
- Resolved many corner cases, like mxw/vim-jsx#180
- Refine syntax group definition, added `jsxElement`, `jsxPuct`, `jsxNamespace` group
- Better support TSX
- Added more test cases
- Much more robust than ever

### CODE REVIEW

**Comment by MaxMEllon** [comment]
Great work... special thanks.

File: after/syntax/javascript.vim [hunk 1]
```diff
@@ -21,24 +20,45 @@ if exists('s:current_syntax')
...
-"          s~~~~~~~~~~~~~~e
-syntax region jsxEscapeJs
-      \ contained
-      \ contains=jsBlock
-      \ start=+{+
-      \ end=++
-      \ extend
+" check the javascript syntax file and cache it
+if !exists('s:js_syntax')
+  let syntax_list = execute('syntax list')
```
Comment (MaxMEllon) [inline after/syntax/javascript.vim]: why use global variable?
this line outside function, so should script local variable, actual: `s:syntax_list`.

Comment (yuezk) [reply after/syntax/javascript.vim]: Changed.

Comment (MaxMEllon) [reply after/syntax/javascript.vim]: Thanks!
